v3: add Webpack 4 upgrade guide information (#1210)
lit/lit.dev: packages/lit-dev-content/site/docs/v3/releases/upgrade.md

Lit 2.x and 3.0 are _interoperable_ – templates, base classes, and directives from one version of Lit will work with those from another.

## Lit is now published as ES2021

Lit 2 was published as ES2019, and Lit 3 is now published as ES2021. ES2021 is supported by modern browsers.
This may be a breaking change for your tooling if it can't parse ES2021, for example in the case of Webpack 4.

### Using Lit 3 with Webpack 4

> Webpack 4's internal parser doesn't support nullish coalescing (`??=`), or optional chaining (`?.`), which are features of ES2021.
> When Webpack 4 parses Lit 3, or other modern packages, it'll throw a `Module parse failed: Unexpected token` on parsing `??=` or `?.`.

Reviewer: ```suggestion
Webpack 4's internal parser doesn't support nullish coalescing (`??`), logical assignment (`??=`), or optional chaining (`?.`), which are syntaxes introduced in ES2021 so will throw a `Module parse failed: Unexpected token` when encountering those.
```

Author: Adding this great change. I think some of these were also technically introduced in ES2020, [`?.` for example](https://262.ecma-international.org/11.0/#prod-OptionalChain). ES2020 also breaks Webpack 4.

> There are two possible solutions:
> 
>  1. Upgrade to Webpack 5, which has updated its internal parser to support ES2021 syntax.
>  2. Transpile Lit 3 to ES2019 syntax using `babel-loader` so Webpack 4 can parse Lit 3.
> 
> Upgrading to Webpack 5 is preferred, otherwise your build system is dictating what target JavaScript you can build.
> Transpiling JavaScript unnecessarily results in a larger bundle with less efficient code.

Reviewer: I think we can make this shorter.
```suggestion
The preferred solution is to upgrade to Webpack 5 which does support parsing these newer JS syntax. However if you are unable to do so, it is possible to use `babel-loader` to transform Lit 3 code to work with Webpack 5.
```

Author: Yes! :D

> As a temporary workaround, to transpile Lit 3 in Webpack 4, install the required babel packages: `npm install --save-dev babel-loader@8 @babel/core @babel/preset-env`.

Reviewer: If going with using the more targeted transform, should install those separately. Also `@babel/core` gets installed via transitive dependency so it doesn't need to be installed separately.
``````suggestion
To transpile Lit 3 in Webpack 4, install the following required babel packages:

```sh
> npm i -D babel-loader@8 \
    @babel/plugin-transform-optional-chaining \
    @babel/plugin-transform-nullish-coalescing-operator \
    @babel/plugin-transform-logical-assignment-operators
```
``````

Author: :D

> And add a new Webpack 4 rule that is similar to the following (you may need to modify it based on your specific project):

Reviewer: ```suggestion
And add a new rule that is similar to the following (you may need to modify it based on your specific project):
```

```js
// In webpack.config.js

module.exports = {
  // ...

  module: {
    rules: [
      // ... your other rules

      // Add a babel-loader rule to downlevel Lit's ES2021 syntax so Webpack 4 can parse it.
      // TODO: Once on Webpack 5, this rule can be deleted.
      {
        test: /\.js$/,
        include: ['@lit', 'lit-element', 'lit-html'].map((p) =>
          path.resolve(__dirname, 'node_modules/' + p)
        ),
        use: {
          loader: 'babel-loader',
          options: {
            presets: [
              [
                '@babel/preset-env',
                {
                  targets: ['last 1 version'],
                },
              ],
            ],

>           },

Author: Legend–and thank you for the precision.

        },
      },
    ],
  }
}
```

> This config transpiles the Lit 3 packages within `node_modules` so Webpack 4 can parse them.
> 

Reviewer: The config snippet above with the comment already says this.
```suggestion
```
